substrate/frame/contracts/src/tests.rs

```diff
@@ -453,6 +454,60 @@ impl Contains<RuntimeCall> for TestFilter {
 	}
 }
 
+parameter_types! {
+	static UploadAccount: <Test as frame_system::Config>::AccountId = ALICE;
+	static InstantiateAccount: <Test as frame_system::Config>::AccountId = ALICE;
+	static RestrictedUpload: bool = false;
+	static RestrictedInstantiation: bool = false;
+}
+
+pub struct EnsureUploadAccount<T>(sp_std::marker::PhantomData<T>);
+impl<T: Config> EnsureOrigin<<T as frame_system::Config>::RuntimeOrigin> for EnsureUploadAccount<T>
+where
+	<T as frame_system::Config>::AccountId: From<AccountId32>,
+{
+	type Success = T::AccountId;
+
+	fn try_origin(o: T::RuntimeOrigin) -> Result<Self::Success, T::RuntimeOrigin> {
+		let who = <frame_system::EnsureSigned<_> as EnsureOrigin<_>>::try_origin(o.clone())?;
+		if RestrictedUpload::get() && who != UploadAccount::get().into() {
+			return Err(o)
+		}
+
+		Ok(who)
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn try_successful_origin() -> Result<T::RuntimeOrigin, ()> {
+		Ok(T::RuntimeOrigin::from(frame_system::RawOrigin::Signed(UploadAccount::get().into())))
+	}
+}
+
+pub struct EnsureInstantiateAccount<T>(sp_std::marker::PhantomData<T>);
+impl<T: Config> EnsureOrigin<<T as frame_system::Config>::RuntimeOrigin>
+	for EnsureInstantiateAccount<T>
+where
+	<T as frame_system::Config>::AccountId: From<AccountId32>,
+{
+	type Success = T::AccountId;
+
+	fn try_origin(o: T::RuntimeOrigin) -> Result<Self::Success, T::RuntimeOrigin> {
+		let who = <frame_system::EnsureSigned<_> as EnsureOrigin<_>>::try_origin(o.clone())?;
+		if RestrictedInstantiation::get() && who != InstantiateAccount::get().into() {
+			return Err(o)
+		}
+
+		Ok(who)
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn try_successful_origin() -> Result<T::RuntimeOrigin, ()> {
+		Ok(T::RuntimeOrigin::from(frame_system::RawOrigin::Signed(
+			InstantiateAccount::get().into(),
+		)))
+	}
+}
+
```

Comment:
Maybe we can just  use a single type here for both Upload & Instantiate Origin

```suggestion
parameter_types! {
	pub static UploadAccount: Option<<Test as frame_system::Config>::AccountId> = None;
	pub static InstantiateAccount: Option<<Test as frame_system::Config>::AccountId> = None;
}

pub struct EnsureAccount<T, A>(sp_std::marker::PhantomData<(T, A)>);
impl<T: Config, A: sp_core::Get<Option<crate::AccountIdOf<T>>>>
	EnsureOrigin<<T as frame_system::Config>::RuntimeOrigin> for EnsureAccount<T, A>
where
	<T as frame_system::Config>::AccountId: From<AccountId32>,
{
	type Success = T::AccountId;

	fn try_origin(o: T::RuntimeOrigin) -> Result<Self::Success, T::RuntimeOrigin> {
		let who = <frame_system::EnsureSigned<_> as EnsureOrigin<_>>::try_origin(o.clone())?;
		if matches!(A::get(), Some(a) if who != a) {
			return Err(o)
		}

		Ok(who)
	}

	#[cfg(feature = "runtime-benchmarks")]
	fn try_successful_origin() -> Result<T::RuntimeOrigin, ()> {
		Err(())
	}
}
```

Then fix up 
```diff
-	type UploadOrigin = EnsureUploadAccount<Self>;
-	type InstantiateOrigin = EnsureInstantiateAccount<Self>;
+	type UploadOrigin = EnsureAccount<Self, UploadAccount>;
+	type InstantiateOrigin = EnsureAccount<Self, InstantiateAccount>;
 	type MaxDebugBufferLen = ConstU32<{ 2 * 1024 * 1024 }>;
 	type RuntimeHoldReason = RuntimeHoldReason;
 	type Migrations = crate::migration::codegen::BenchMigrations;
@@ -6007,7 +5980,7 @@ fn root_cannot_instantiate() {
 #[test]
 fn only_upload_origin_can_upload() {
 	let (wasm, _) = compile_module::<Test>("dummy").unwrap();
-	RestrictedUpload::set(true);
+	UploadAccount::set(Some(ALICE));
```